[Quant][PT2E]Fix the maxpool2d input observer didn't insert after QuantizationAnnotation API

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102516
* #98826
* #104075
* #104074
* __->__ #101941

**Summary**
The previous UT has been broken accidently, since the output of conv2d node has been annotated by mistake.
Re-enable these UTs for case:

- Single `conv2d` node, if we don't annotate the output node of `conv2d`. There should be no fake quant at conv2d's output.
-  For `conv2d-maxpool` pattern, `maxpool` should has fake quant inserted at input and output node since we annotate these nodes.

**Test Plan**
```
python -m pytest test_quantize_pt2e.py -k test_wo_annotate_conv_output_quantizer
python -m pytest test_quantize_pt2e.py -k test_max_pool2d_quantizer
```


cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

leslie-fang-intel committed:
Fix the maxpool2d input observer didn't insert after QuantizationAnnotation API

[ghstack-poisoned]

leslie-fang-intel commented:
Hi @jerryzh168, could you help to take a look of this PR? PR https://github.com/pytorch/pytorch/pull/101708 has broken the test recipe in UT of `test_max_pool2d_quantizer` and the UT modification looks not reasonable to me. In the recipe tested by this UT, we should not add `output_qspec` to the conv node.

jerryzh168 commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -523,7 +523,7 @@ def _get_output_act_obs_or_fq_ctr(
     """
     assert isinstance(arg, Node)
-    if "quantization_annotation" in arg.meta:
+    if "quantization_annotation" in arg.meta and arg.meta["quantization_annotation"].output_qspec is not None:
```
is this needed?

leslie-fang-intel replied:
Yes, I think so. Otherwise a `None` object is return by `_get_output_act_obs_or_fq_ctr` and it fails the check of `_needs_obs_or_fq` in https://github.com/pytorch/pytorch/blob/8243abc84ada1a7c8153d7257b60ba4ffdc7b7cb/torch/ao/quantization/_pt2e/prepare.py#L66-L74. So there will be no observer inserted at input of `maxpool2d`.

jerryzh168 replied:
I see, this logic is changed again in the next PR (not yet in oss), please wait a bit

leslie-fang-intel replied:
Thanks, will wait for next PRs. Please let me know when to do the rebase.

jerryzh168 commented:
> Hi @jerryzh168, could you help to take a look of this PR? PR #101708 has broken the test recipe in UT of `test_max_pool2d_quantizer` and the UT modification looks not reasonable to me. In the recipe tested by this UT, we should not add `output_qspec` to the conv node.

oh I see, I think it's a mistake during refactor, could you add a check for this?

leslie-fang-intel commented:
> > Hi @jerryzh168, could you help to take a look of this PR? PR #101708 has broken the test recipe in UT of `test_max_pool2d_quantizer` and the UT modification looks not reasonable to me. In the recipe tested by this UT, we should not add `output_qspec` to the conv node.
> 
> oh I see, I think it's a mistake during refactor, could you add a check for this?

Great suggestion. I have added a check in this UT to ensure the `output_qspec` for `conv` node is None.

leslie-fang-intel committed:
Update on "Fix the maxpool2d input observer didn't insert after QuantizationAnnotation API"


**Summary**
After `QuantizationAnnotation API` merged in PR https://github.com/pytorch/pytorch/pull/101708/ , the previous PR https://github.com/pytorch/pytorch/pull/100961 to fix issue https://github.com/pytorch/pytorch/issues/100959 is broken. In this PR, we fixed this regression. When `arg.meta["quantization_annotation"].output_qspec` is `None`, we shouldn't use it as `output_act_obs_or_fq_ctr`.

**Test Plan**
```
python -m pytest test_quantize_pt2e.py -k test_max_pool2d_quantizer
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E]Fix the maxpool2d input observer didn't insert after QuantizationAnnotation API"


**Summary**
After `QuantizationAnnotation API` merged in PR https://github.com/pytorch/pytorch/pull/101708/ , the previous PR https://github.com/pytorch/pytorch/pull/100961 to fix issue https://github.com/pytorch/pytorch/issues/100959 is broken. In this PR, we fixed this regression. When `arg.meta["quantization_annotation"].output_qspec` is `None`, we shouldn't use it as `output_act_obs_or_fq_ctr`.

**Test Plan**
```
python -m pytest test_quantize_pt2e.py -k test_max_pool2d_quantizer
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E]Fix the maxpool2d input observer didn't insert after QuantizationAnnotation API"


**Summary**
After `QuantizationAnnotation API` merged in PR https://github.com/pytorch/pytorch/pull/101708/ , the previous PR https://github.com/pytorch/pytorch/pull/100961 to fix issue https://github.com/pytorch/pytorch/issues/100959 is broken. In this PR, we fixed this regression. When `arg.meta["quantization_annotation"].output_qspec` is `None`, we shouldn't use it as `output_act_obs_or_fq_ctr`.

**Test Plan**
```
python -m pytest test_quantize_pt2e.py -k test_max_pool2d_quantizer
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E]Fix the maxpool2d input observer didn't insert after QuantizationAnnotation API"


**Summary**
After `QuantizationAnnotation API` merged in PR https://github.com/pytorch/pytorch/pull/101708/ , the previous PR https://github.com/pytorch/pytorch/pull/100961 to fix issue https://github.com/pytorch/pytorch/issues/100959 is broken. In this PR, we fixed this regression. When `arg.meta["quantization_annotation"].output_qspec` is `None`, we shouldn't use it as `output_act_obs_or_fq_ctr`.

**Test Plan**
```
python -m pytest test_quantize_pt2e.py -k test_max_pool2d_quantizer
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel commented:
Hi @jerryzh168, the changes in PR https://github.com/pytorch/pytorch/pull/102184 has fixed this issue. So I only keep the change revert of UT in this PR. Please help to take a look.

jerryzh168 commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -231,6 +230,13 @@ def get_supported_operators(cls) -> List[OperatorConfig]:
             aten_graph=True,
         )
         m = prepare_pt2e_quantizer(m, BackendAQuantizer())
+        for node in m.graph.nodes:
```
can we check this in a different way.. it feels like this is exposing too much implementation detail, maybe just check the number of observers inserted?

leslie-fang-intel replied:
Thanks for the suggestion and changed.

leslie-fang-intel replied:
@jerryzh168  Could you take a look of this PR again? The previous comments have been adjusted.

leslie-fang-intel committed:
Update on "[Quant][PT2E]Fix the maxpool2d input observer didn't insert after QuantizationAnnotation API"


**Summary**
After `QuantizationAnnotation API` merged in PR https://github.com/pytorch/pytorch/pull/101708/ , the previous PR https://github.com/pytorch/pytorch/pull/100961 to fix issue https://github.com/pytorch/pytorch/issues/100959 is broken. In this PR, we fixed this regression. When `arg.meta["quantization_annotation"].output_qspec` is `None`, we shouldn't use it as `output_act_obs_or_fq_ctr`.

**Test Plan**
```
python -m pytest test_quantize_pt2e.py -k test_max_pool2d_quantizer
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E]Fix the maxpool2d input observer didn't insert after QuantizationAnnotation API"


**Summary**
After `QuantizationAnnotation API` merged in PR https://github.com/pytorch/pytorch/pull/101708/ , the previous PR https://github.com/pytorch/pytorch/pull/100961 to fix issue https://github.com/pytorch/pytorch/issues/100959 is broken. In this PR, we fixed this regression. When `arg.meta["quantization_annotation"].output_qspec` is `None`, we shouldn't use it as `output_act_obs_or_fq_ctr`.

**Test Plan**
```
python -m pytest test_quantize_pt2e.py -k test_max_pool2d_quantizer
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E]Fix the maxpool2d input observer didn't insert after QuantizationAnnotation API"


**Summary**
After `QuantizationAnnotation API` merged in PR https://github.com/pytorch/pytorch/pull/101708/ , the previous PR https://github.com/pytorch/pytorch/pull/100961 to fix issue https://github.com/pytorch/pytorch/issues/100959 is broken. In this PR, we fixed this regression. When `arg.meta["quantization_annotation"].output_qspec` is `None`, we shouldn't use it as `output_act_obs_or_fq_ctr`.

**Test Plan**
```
python -m pytest test_quantize_pt2e.py -k test_max_pool2d_quantizer
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel commented:
@jerryzh168 Could you help to take a look of this PR again? The previous comments have been adjusted.

leslie-fang-intel committed:
Update on "[Quant][PT2E]Fix the maxpool2d input observer didn't insert after QuantizationAnnotation API"


**Summary**
After `QuantizationAnnotation API` merged in PR https://github.com/pytorch/pytorch/pull/101708/ , the previous PR https://github.com/pytorch/pytorch/pull/100961 to fix issue https://github.com/pytorch/pytorch/issues/100959 is broken. In this PR, we fixed this regression. When `arg.meta["quantization_annotation"].output_qspec` is `None`, we shouldn't use it as `output_act_obs_or_fq_ctr`.

**Test Plan**
```
python -m pytest test_quantize_pt2e.py -k test_max_pool2d_quantizer
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E]Fix the maxpool2d input observer didn't insert after QuantizationAnnotation API"


**Summary**
After `QuantizationAnnotation API` merged in PR https://github.com/pytorch/pytorch/pull/101708/ , the previous PR https://github.com/pytorch/pytorch/pull/100961 to fix issue https://github.com/pytorch/pytorch/issues/100959 is broken. In this PR, we fixed this regression. When `arg.meta["quantization_annotation"].output_qspec` is `None`, we shouldn't use it as `output_act_obs_or_fq_ctr`.

**Test Plan**
```
python -m pytest test_quantize_pt2e.py -k test_max_pool2d_quantizer
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E]Fix the maxpool2d input observer didn't insert after QuantizationAnnotation API"


**Summary**
After `QuantizationAnnotation API` merged in PR https://github.com/pytorch/pytorch/pull/101708/ , the previous PR https://github.com/pytorch/pytorch/pull/100961 to fix issue https://github.com/pytorch/pytorch/issues/100959 is broken. In this PR, we fixed this regression. When `arg.meta["quantization_annotation"].output_qspec` is `None`, we shouldn't use it as `output_act_obs_or_fq_ctr`.

**Test Plan**
```
python -m pytest test_quantize_pt2e.py -k test_max_pool2d_quantizer
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E]Fix the maxpool2d input observer didn't insert after QuantizationAnnotation API"


**Summary**
The previous UT has been broken accidently, since the output of conv2d node has been annotated by mistake.
Re-enable these UTs for case:

- Single `conv2d` node, if we don't annotate the output node of `conv2d`. There should be no fake quant at conv2d's output.
-  For `conv2d-maxpool` pattern, `maxpool` should has fake quant inserted at input and output node since we annotate these nodes.

**Test Plan**
```
python -m pytest test_quantize_pt2e.py -k test_max_pool2d_quantizer
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E]Fix the maxpool2d input observer didn't insert after QuantizationAnnotation API"


**Summary**
The previous UT has been broken accidently, since the output of conv2d node has been annotated by mistake.
Re-enable these UTs for case:

- Single `conv2d` node, if we don't annotate the output node of `conv2d`. There should be no fake quant at conv2d's output.
-  For `conv2d-maxpool` pattern, `maxpool` should has fake quant inserted at input and output node since we annotate these nodes.

**Test Plan**
```
python -m pytest test_quantize_pt2e.py -k test_max_pool2d_quantizer
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -574,18 +576,14 @@ def annotate(self, model: torch.fx.GraphModule) -> torch.fx.GraphModule:
                         and node.target == torch.ops.aten.convolution.default
                     ):
                         input_act = node.args[0]
-                        assert isinstance(input_act, Node)
```
Since this quantizer is only used for test case of `conv->maxpool`, I think removing it makes the code clean. Anyway, I add it back. Please help to take a look again.

jerryzh168 commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -621,29 +622,50 @@ def validate(self, model: torch.fx.GraphModule) -> None:
-        self._test_quantizer(
-            m, example_inputs, BackendAQuantizer(), node_occurrence, node_list
-        )
+        for use_maxpool2d in [True, False]:
```
maybe just split this to two tests?

leslie-fang-intel replied:
Thanks for the suggestion. Split into two tests of `test_wo_annotate_conv_output_quantizer` and `test_max_pool2d_quantizer`. Please kindly help to take a look again @jerryzh168.

leslie-fang-intel committed:
Update on "[Quant][PT2E]Fix the maxpool2d input observer didn't insert after QuantizationAnnotation API"


**Summary**
The previous UT has been broken accidently, since the output of conv2d node has been annotated by mistake.
Re-enable these UTs for case:

- Single `conv2d` node, if we don't annotate the output node of `conv2d`. There should be no fake quant at conv2d's output.
-  For `conv2d-maxpool` pattern, `maxpool` should has fake quant inserted at input and output node since we annotate these nodes.

**Test Plan**
```
python -m pytest test_quantize_pt2e.py -k test_wo_annotate_conv_output_quantizer
python -m pytest test_quantize_pt2e.py -k test_max_pool2d_quantizer
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E]Fix the maxpool2d input observer didn't insert after QuantizationAnnotation API"


**Summary**
The previous UT has been broken accidently, since the output of conv2d node has been annotated by mistake.
Re-enable these UTs for case:

- Single `conv2d` node, if we don't annotate the output node of `conv2d`. There should be no fake quant at conv2d's output.
-  For `conv2d-maxpool` pattern, `maxpool` should has fake quant inserted at input and output node since we annotate these nodes.

**Test Plan**
```
python -m pytest test_quantize_pt2e.py -k test_wo_annotate_conv_output_quantizer
python -m pytest test_quantize_pt2e.py -k test_max_pool2d_quantizer
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E]Fix the maxpool2d input observer didn't insert after QuantizationAnnotation API"


**Summary**
The previous UT has been broken accidently, since the output of conv2d node has been annotated by mistake.
Re-enable these UTs for case:

- Single `conv2d` node, if we don't annotate the output node of `conv2d`. There should be no fake quant at conv2d's output.
-  For `conv2d-maxpool` pattern, `maxpool` should has fake quant inserted at input and output node since we annotate these nodes.

**Test Plan**
```
python -m pytest test_quantize_pt2e.py -k test_wo_annotate_conv_output_quantizer
python -m pytest test_quantize_pt2e.py -k test_max_pool2d_quantizer
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E]Fix the maxpool2d input observer didn't insert after QuantizationAnnotation API"


**Summary**
The previous UT has been broken accidently, since the output of conv2d node has been annotated by mistake.
Re-enable these UTs for case:

- Single `conv2d` node, if we don't annotate the output node of `conv2d`. There should be no fake quant at conv2d's output.
-  For `conv2d-maxpool` pattern, `maxpool` should has fake quant inserted at input and output node since we annotate these nodes.

**Test Plan**
```
python -m pytest test_quantize_pt2e.py -k test_wo_annotate_conv_output_quantizer
python -m pytest test_quantize_pt2e.py -k test_max_pool2d_quantizer
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E]Fix the maxpool2d input observer didn't insert after QuantizationAnnotation API"


**Summary**
The previous UT has been broken accidently, since the output of conv2d node has been annotated by mistake.
Re-enable these UTs for case:

- Single `conv2d` node, if we don't annotate the output node of `conv2d`. There should be no fake quant at conv2d's output.
-  For `conv2d-maxpool` pattern, `maxpool` should has fake quant inserted at input and output node since we annotate these nodes.

**Test Plan**
```
python -m pytest test_quantize_pt2e.py -k test_wo_annotate_conv_output_quantizer
python -m pytest test_quantize_pt2e.py -k test_max_pool2d_quantizer
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge

leslie-fang-intel commented:
@pytorchbot merge
